File: ts/10 - Backtracking/78.Subsets.ts
/* 
https://leetcode.com/problems/subsets/description/
*/

/* ------------------------------------------------------------------------------- */

// function subsets(nums: number[]): number[][] {
//     const result: number[][] = [];
//     const subset: number[] = [];

//     function dfs(idx: number) {
//         result.push([...subset]);

//         for (let i = idx; i < nums.length; i++) {
//             subset.push(nums[i]);
//             dfs(i + 1);
//             subset.pop();
//         }
//     }

//     dfs(0);
//     return result;
// }

/* ------------------------------------------------------------------------------- */

// function subsets(nums: number[]): number[][] {
//   const result: number[][] = []
//   const subset: number[] = []

//   function dfs(idx: number) {
//     if (idx > nums.length - 1) return result.push([...subset])

//     subset.push(nums[idx])
//     dfs(idx + 1)
//     subset.pop()
//     dfs(idx + 1)
//   }

//   dfs(0)

//   return result
// }

/* ------------------------------------------------------------------------------- */

/**
 *  每一個數字都有兩種選擇，放進 subset 或不放進 subset
 *                  [1]                      []
 *          [1,2]          [1]        [2]         []
 *      [1,2,3]  [1,2] [1,3]  [1]  [2,3]  [2]   [3]  []
 */

/*
T.C.: O(N * 2^N)
- 每個數字的兩種選擇共 2^N 種可能。
- 在每個回溯步驟中，將元素添加到 subset 中，這需要 O(N) 時間。

S.C.: O(N)
- 除了輸入和輸出，額外的空間主要用於遞迴呼叫的 call stack 和 subset，最壞情況下，call stack 的深度是 N。
- subset 的隨著遞迴的進行而變化，最大大小不會超過輸入的陣列長度 N。
*/

function subsets(nums) {
  let res: Array<Array<number>> = [[]]

  for (let num of nums) {
    let size = res.length
    for (let i = 0; i < size; i++) {
      let subset = res[i].slice()
      subset.push(num)
      res.push(subset)
    }
  }

  return res
}

const nums = [1, 2, 3]
console.log(subsets(nums))

File: python/10 - Backtracking/78.Subsets.py
from typing import List

# recursion
# class Solution:
#     def subsets(self, nums: List[int]) -> List[List[int]]:
#         subset = []
#         res = []

#         def dfs(idx: int):
#             if idx >= len(nums):
#                 return res.append(subset.copy())
#             subset.append(nums[idx])
#             dfs(idx + 1)
#             subset.pop()
#             dfs(idx + 1)

#         dfs(0)
#         return res


"""
T.C: O(N × 2^N)
- Total subsets generated: 2^N (each element has 2 choices)
- Copying subset at each leaf: O(N) time using subset.copy()
- Total operations: 2^N subsets × N copying time = O(N × 2^N)
- Recursive calls: 2^N total calls, each doing O(1) work except copying

S.C: O(N)
- Recursion call stack depth: O(N) maximum depth
- Temporary subset list: O(N) maximum size when all elements included
- Output space not counted in auxiliary space complexity
- No additional data structures beyond the recursive stack and subset
"""


# iteration
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        res = [[]]

        for num in nums:
            res += [subset + [num] for subset in res]

        return res


"""
  T.C: O(n × 2^n)
  - Outer loop runs n times (for each number)
  - At iteration i, we have 2^i existing subsets
  - Each iteration creates 2^i new subsets by copying and
  extending
  - Total operations: Σ(i=0 to n-1) 2^i = 2^n - 1 ≈ O(2^n)
  - Each subset creation involves list concatenation:
  O(average subset length) = O(n)
  - Overall: O(n × 2^n)

  S.C: O(2^n)
  - Result stores 2^n subsets
  - Each subset averages n/2 elements
  - Auxiliary space for list comprehension: O(2^i) at
  iteration i
  - Overall dominated by result storage: O(2^n)
"""


# Test cases
if __name__ == "__main__":
    solution = Solution()

    # Test case 1: Basic example
    nums1 = [1, 2, 3]
    result1 = solution.subsets(nums1)
    print(f"Input: {nums1}")
    print(f"Output: {result1}")
    print("Expected: [[], [1], [2], [1,2], [3], [1,3], [2,3], [1,2,3]]")
    print()

    # Test case 2: Single element
    nums2 = [0]
    result2 = solution.subsets(nums2)
    print(f"Input: {nums2}")
    print(f"Output: {result2}")
    print("Expected: [[], [0]]")
    print()

    # Test case 3: Two elements
    nums3 = [1, 2]
    result3 = solution.subsets(nums3)
    print(f"Input: {nums3}")
    print(f"Output: {result3}")
    print("Expected: [[], [1], [2], [1,2]]")
    print()

    # Test case 4: Larger array
    nums4 = [1, 2, 3, 4]
    result4 = solution.subsets(nums4)
    print(f"Input: {nums4}")
    print(f"Output length: {len(result4)} (should be 16)")
    print("First few subsets:", result4[:8])
